File: Assets/Code/Logic/UI/Description/View/CommonItemDesView.cs
using UnityEngine;
using System.Collections;
using Logic.Item.Model;
using UnityEngine.UI;
using Common.Localization;
using Logic.UI.CommonItem.View;
using System.Collections.Generic;
using Logic.Enums;
using Logic.UI.GoodsJump.Model;

namespace Logic.UI.Description.View
{
	public class CommonItemDesView : MonoBehaviour {
		
		public const string PREFAB_PATH = "ui/description/common_item_description_view";
		
		public static CommonItemDesView Open(ItemInfo data, Vector3 pos,Vector2 sizeDelta)
		{
			CommonItemDesView view = UIMgr.instance.Open<CommonItemDesView>(PREFAB_PATH, EUISortingLayer.Tips);
			view.SetData(data,pos,sizeDelta);

			return view;
		}

		public Text textTitle;
		public Transform iconRoot;
		public Text textDes;
		public Text textFrom;
		public Text textUse;
		public RectTransform rootPanel;
		public RectTransform bottomRoot;
		private float _defaultContentLineHeight = 25;

		private int _defaultBorderX = 10;

		private ItemInfo _itemInfo;
		private Vector3 _worldPos;
		private Vector2 _sizeDelta ;
		private Vector2 _originSizeDelta;

		void Awake()
		{
			_originSizeDelta = rootPanel.sizeDelta;
			_defaultContentLineHeight = textUse.preferredHeight;
		}

		public void SetData(ItemInfo data,Vector3 worldPos,Vector2 sizeDelta)
		{

			_itemInfo = data;
			_worldPos = worldPos;
			_sizeDelta = sizeDelta;
			StartCoroutine(RefreshCoroutine());

		}
		public void SetPivot(Vector2 pivot)
		{
			rootPanel.pivot = pivot;
		}
		public IEnumerator RefreshCoroutine()
		{
			rootPanel.gameObject.SetActive(false);
			yield return null;

			textTitle.text = Localization.Get(_itemInfo.itemData.name);
			textDes.text = Localization.Get(_itemInfo.itemData.des);
			DropMessageData data = DropMessageData.GetDropMsgDataByResData((int)BaseResType.Item,_itemInfo.itemData.id);
			textFrom.text = data == null ? "" : Localization.Get( data.des);
			CommonItemIcon icon =  CommonItemIcon.Create(iconRoot);
			icon.SetItemInfo(new ItemInfo(0,_itemInfo.itemData.id,0));
			icon.GetComponent<ItemDesButton>().enabled = false;
			RefreshSpecialUse();
			float useDeltaHeight = (string.IsNullOrEmpty(textUse.text) ? 0 : textUse.preferredHeight)-_defaultContentLineHeight;
			float fromDeltaHeight = (string.IsNullOrEmpty(textFrom.text) ? 0 : textFrom.preferredHeight)-_defaultContentLineHeight;
			float deltaHeight = useDeltaHeight + fromDeltaHeight;
			textFrom.transform.localPosition += new Vector3(0, useDeltaHeight);
			rootPanel.sizeDelta = new Vector2(_originSizeDelta.x,_originSizeDelta.y+deltaHeight);

			float screenHalfHeight = UIMgr.instance.designResolution.y/2;
			Vector3 localPosition = transform.InverseTransformPoint(_worldPos);
			float x = 0f;
			float y = localPosition.y+rootPanel.sizeDelta.y/2;
			if(localPosition.x>0)
			{
				x = localPosition.x-_sizeDelta.x/2-rootPanel.sizeDelta.x/2-_defaultBorderX;
			}else{
				x = localPosition.x+_sizeDelta.x/2+rootPanel.sizeDelta.x/2+_defaultBorderX;
			}
			if(y < rootPanel.sizeDelta.y-screenHalfHeight)
			{
				y = rootPanel.sizeDelta.y-screenHalfHeight;
			}
			if(y > screenHalfHeight)
			{
				y = screenHalfHeight;
			}
			localPosition = new Vector3(x,y);
			rootPanel.localPosition = localPosition;
			rootPanel.gameObject.SetActive(true);
		}
		private void RefreshSpecialUse()
		{
			List<RoleType> heroTypeList = _itemInfo.itemData.heroTypeList;
			int count = heroTypeList.Count ;
			if(count == 0 )
			{
				textUse.gameObject.SetActive(false);
			}else
			{
				
				textUse.gameObject.SetActive(true);
				string typeName = string.Empty;
				for(int i = 0;i<count;i++)
				{
					if(i == 0)
						typeName = UIUtil.GetRoleTypeName((RoleType)heroTypeList[i]);
					else 
					typeName = string.Format("{0}、{1}",typeName,UIUtil.GetRoleTypeName((RoleType)heroTypeList[i]));
				}
				textUse.text = string.Format(Localization.Get("ui.common_des_view.use"),typeName);
			}
		}

		public void Close()
		{
			UIMgr.instance.Close(PREFAB_PATH);
		}
	}
}

File: Assets/ToLua/Lua/user/ui/description/common_item_desview.lua
local tips={}
local PREFAB_PATH = "ui/description/common_item_description_view_lua"

local _defaultContentLineHeight=25
local _defaultBorderX=10
local textTitle
local iconRoot
local textDes
local textFrom
local textUse
local rootPanel
local bottomRoot
local itemInfo
local worldPos
local sizeDelta 
local originSizeDelta

local function Start( ... )
end

function tips.Open(info,pos,sizeDelta)
  local t = Logic.UI.UIMgr.instance:Open(PREFAB_PATH,EUISortingLayer.Tips,0).transform
  textTitle=t:FindChild('core/frame/text_title'):GetComponent(typeof(Text))
  iconRoot=t:FindChild('core/frame/icon_root')
  textDes=t:FindChild('core/frame/text_des'):GetComponent(typeof(Text))
  textFrom=t:FindChild('core/frame/bottom_root/text_from'):GetComponent(typeof(Text))
  textUse=t:FindChild('core/frame/bottom_root/text_use'):GetComponent(typeof(Text))
  rootPanel=t:FindChild('core/frame'):GetComponent(typeof(RectTransform))
  bottomRoot=t:FindChild('core/frame/bottom_root'):GetComponent(typeof(RectTransform))
  t:FindChild('core/btn_full_close'):GetComponent(typeof(Button)).onClick:AddListener(tips.Close)
  t:FindChild('core/frame'):GetComponent(typeof(Button)).onClick:AddListener(tips.Close)
  originSizeDelta = rootPanel.sizeDelta
  _defaultContentLineHeight = textUse.preferredHeight

  tips.transform=t
  tips.SetData(info,pos,sizeDelta)
end

function tips.SetData(info,pos,sizedelta)
  itemInfo = info
  worldPos = pos
  sizeDelta = sizedelta
  tips.co=coroutine.start(tips.RefreshCoroutine)
end

function tips.RefreshCoroutine()
  rootPanel.gameObject:SetActive(false)
  coroutine.step()

  textTitle.text = LocalizationController.instance:Get(itemInfo.itemData.name)
  textDes.text = LocalizationController.instance:Get(itemInfo.itemData.des)

  textFrom.text = LuaCsTransfer.GetDropDataDes(BaseResType.Item,itemInfo.itemData.id)

  local icon = require('ui/common_icon/common_item_icon').New(iconRoot)
  icon:SetItemInfo(require('ui/item/model/item_info').New(0,itemInfo.itemData.id,0))

  --maybe not necessary
  --icon.GetComponent<ItemDesButton>().enabled = false

  tips.RefreshSpecialUse()
  local useDeltaHeight = 0
  if not textUse.text or textUse.text == '' then
    useDeltaHeight=0-_defaultContentLineHeight
  else
    useDeltaHeight=textUse.preferredHeight-_defaultContentLineHeight
  end
  local fromDeltaHeight = 0
  if not textFrom.text or textFrom.text == '' then
    fromDeltaHeight=0-_defaultContentLineHeight
  else
    fromDeltaHeight=textFrom.preferredHeight-_defaultContentLineHeight
  end

  local deltaHeight = useDeltaHeight + fromDeltaHeight
  textFrom.transform.localPosition =textFrom.transform.localPosition+ Vector3(0, useDeltaHeight)
  rootPanel.sizeDelta = Vector2(originSizeDelta.x,originSizeDelta.y+deltaHeight)

  local screenHalfHeight = UIMgr.instance.designResolution.y/2
  local localPosition = tips.transform:InverseTransformPoint(worldPos)
  local x = 0
  local y = localPosition.y
  if localPosition.x>0 then
    x = localPosition.x-sizeDelta.x/2-rootPanel.sizeDelta.x/2-_defaultBorderX
  else
    x = localPosition.x+sizeDelta.x/2+rootPanel.sizeDelta.x/2+_defaultBorderX
  end

  if localPosition.y<rootPanel.sizeDelta.y/2-screenHalfHeight then
    y = rootPanel.sizeDelta.y/2-screenHalfHeight
  end

  if localPosition.y>screenHalfHeight-rootPanel.sizeDelta.y/2 then
    y = screenHalfHeight - rootPanel.sizeDelta.y/2
  end
  localPosition = Vector3(x,y)
  rootPanel.localPosition = localPosition
  rootPanel.gameObject:SetActive(true)
end

function tips.RefreshSpecialUse()
  local heroTypeList = itemInfo.itemData.hero_type
  local count = #heroTypeList
  if count == 0 then
    textUse.gameObject:SetActive(false)
  else
    local typeName = ''
    for i = 1,count do
      if i == 1 then
        typeName = ui_util.GetRoleTypeName(heroTypeList[i])
      else 
        typeName = string.format("%s、%s",typeName,ui_util.GetRoleTypeName(heroTypeList[i]))
      end
    end
    textUse.text = string.format(LocalizationController.instance:Get("ui.common_des_view.use_lua"),typeName)
  end
  textUse.gameObject:SetActive(true)
end

function tips.Close()
  coroutine.stop(tips.co)
  UIMgr.instance:Close(PREFAB_PATH)
end

Start()
return tips